http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -93,7 +93,13 @@ makeConnection r w c = do
     -- already closed connection.
     closedVar <- newIORef False
 
-    _ <- mkWeakIORef istack c
+    let close = do
+          closed <- readIORef closedVar
```

Comment:
I'm not certain it's necessary, but to avoid a potential race condition, what about using `atomicModifyIORef` to read the closed state and immediately write back `True`?